io/runes/runes_test.go

```diff
@@ -0,0 +1,177 @@
+
+var testcases = []testcase{
+	{
+		name:   "empty input",
+		input:  "",
+		repeat: 1,
+	},
+	{
+		name:   "invalid code point",
+		input:  string([]byte{0x80}),
+		repeat: 1,
+		want: testResult{
+			err: fmt.Errorf("invalid rune at offset 0"),
+		},
+	},
+	{
+		name:   "simple ascii",
+		input:  "test",
+		repeat: 1,
+	},
+	{
+		name:   "single rune - 2 bytes",
+		input:  "ί",
+		repeat: 1,
+	},
+	{
+		name:   "1 ascii - start with multibyte",
+		input:  "ίA",
+		repeat: 1,
+	},
+	{
+		name:   "1 ascii - end with multibyte",
+		input:  "Aί",
+		repeat: 1,
+	},
+	{
+		name:   "single rune - 3 bytes",
+		input:  "ࠀ",
+		repeat: 1,
+	},
+	{
+		name:   "decoding ࠀࠆࠉࠌ",
+		input:  "ࠀࠆࠉࠌ",
+		repeat: 1,
+	},
+	{
+		name:   "decoding mixed 3-byte and ASCII ࠀࠆࠉASCIIࠌ",
+		input:  "ࠀࠆࠉASCIIࠌ",
+		repeat: 1,
+	},
+	{
+		name:   "single rune - 4 bytes",
+		input:  "𒁘",
+		repeat: 1,
+	},
+	{
+		name:   "decoding 𓁺𒂞𒆙𒈙𒌦𓁙",
+		input:  "𓁺𒂞𒆙𒈙𒌦𓁙",
+		repeat: 1,
+	},
+	{
+		name:   "decoding mixed ASCII and 4-byte - 𓁺𒂞𒆙ASCII𒈙𒌦𓁙",
+		input:  "𓁺𒂞𒆙ASCII𒈙𒌦𓁙",
+		repeat: 1,
+	},
+	{
+		name:   "decoding " + socraticParadox,
+		input:  socraticParadox,
+		repeat: 1,
+	},
+	{
+		name:   "decoding multibyte code points + ascii",
+		input:  "Εν οίδα ότι TESTE ουδέν οίδα",
+		repeat: 1,
+	},
+	{
+		name:   "ascii with newline",
+		input:  "test\n",
+		repeat: 1024,
+	},
+	{
+		name:   "multibyte - 1024 bytes",
+		input:  socraticParadox,
+		repeat: 1024,
+	},
+	{
+		name:   "multibyte - 1025 bytes",
+		input:  socraticParadox,
+		repeat: 1025,
+	},
+	{
+		name:   "multibyte - 2048 bytes",
+		input:  socraticParadox,
+		repeat: 2048,
+	},
+	{
+		name:   "multibyte - 2049 bytes",
+		input:  socraticParadox,
+		repeat: 2049,
+	},
+	{
+		name:   "invalid rune sequence",
+		input:  string([]byte{206, 149, 206, 206, 206}),
+		repeat: 1,
+		want: testResult{
+			err:     fmt.Errorf("invalid rune at offset 2"),
+			partial: []rune{'Ε'},
+		},
+	},
+}
+
+func TestUnicodeDecoder(t *testing.T) {
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			inputReader := iotest.NewRepeatReader(
+				strings.NewReader(tc.input),
+				tc.repeat)
+
+			data, err := io.ReadAll(inputReader)
+			assert.NoError(t, err, "reading repeated string")
+
+			reader := runes.NewUnicodeReader(bytes.NewBuffer(data))
+
+			got, err := runes.ReadAll(reader)
+			assert.EqualErrs(t, tc.want.err, err, "readAll error mismatch")
+
+			var expected []rune
+
+			if tc.want.err == nil {
+				repeater := iotest.NewRepeatReader(strings.NewReader(tc.input),
+					tc.repeat)
+
+				expectedBytes, err := io.ReadAll(repeater)
+				assert.NoError(t, err, "repeating expected")
+
+				expected = []rune(string(expectedBytes))
+			} else {
+				expected = tc.want.partial
+			}
+
+			assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch")
+
+			for i, r := range expected {
+				if r != got[i] {
+					t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
```

Comment:
```suggestion
					t.Errorf("want[%[1]d = %[1]c] but got[%[2]d = %[2]c]", r, got[i])
```

Reply:
I don't like this formatting.... =(

Reply:
uaheuaheuhaue yeah it is fair enough...not in love with it either, but I also find funny to pass repeated parameters to formatting, so in the end I dislike both :laughing: . Wanted to see what you would think about it =P